packages/flutter_tools/lib/src/flutter_plugins.dart

```diff
@@ -1157,127 +1157,242 @@ List<PluginInterfaceResolution> resolvePlatformImplementation(
-          );
-          hasResolutionError = true;
-        } else {
-          pluginResolution.add(directDependencies.first);
+    // Now resolve all the possible resolutions to a single option for each
+    // plugin, or throw if that's not possible.
+    for (final MapEntry<String, List<Plugin>> implCandidatesEntry in pluginImplCandidates.entries) {
+      try {
+        final Plugin? resolution = _resolveImplementationOfPlugin(
+          platformKey: platformKey,
+          pluginName: implCandidatesEntry.key,
+          candidates: implCandidatesEntry.value,
+          defaultPackageName: defaultImplementations[implCandidatesEntry.key],
+        );
+        if (resolution != null) {
+          pluginResolution[implCandidatesEntry.key] = resolution;
         }
-        continue;
-      }
-      // Next, defer to the default implementation if there is one.
-      final String? defaultPackageName = defaultImplementations[entry.key];
-      if (defaultPackageName != null) {
-        final int defaultIndex = candidates
-            .indexWhere((Plugin plugin) => plugin.name == defaultPackageName);
-        if (defaultIndex != -1) {
-          pluginResolution.add(candidates[defaultIndex]);
-          continue;
+      } on ToolExit catch (e) {
+        if (e.message != null) {
+          globals.printError(e.message!);
         }
-      }
-      // Otherwise, require an explicit choice.
-      if (candidates.length > 1) {
-        globals.printError(
-            'Plugin ${entry.key}:$platformKey has multiple possible implementations:\n'
-                '${candidates.map((Plugin plugin) => '  ${plugin.name}\n').join()}'
-                'To fix this issue, add one of these dependencies to pubspec.yaml.\n'
-        );
         hasResolutionError = true;
-        continue;
       }
     }
 
-    finalResolution.addAll(
-      pluginResolution.map((Plugin plugin) =>
+    pluginResolutions.addAll(
+      pluginResolution.values.map((Plugin plugin) =>
           PluginInterfaceResolution(plugin: plugin, platform: platformKey)),
     );
   }
+  if (hasPluginPubspecError) {
+    throwToolExit('Please resolve the plugin pubspec errors');
+  }
   if (hasResolutionError) {
     throwToolExit('Please resolve the plugin implementation selection errors');
   }
-  return finalResolution;
+  return pluginResolutions;
+}
+
+/// Validates conflicting plugin parameters in pubspec, such as
+/// `dartPluginClass`, `default_package` and `implements`.
+void _validatePlugin(Plugin plugin, String platformKey) {
+  final String? implementsPackage = plugin.implementsPackage;
+  final String? defaultImplPluginName = plugin.defaultPackagePlatforms[platformKey];
+
+  if (plugin.name == implementsPackage &&
+      plugin.name == defaultImplPluginName) {
+    // Allow self implementing and self as platform default.
+    return;
+  }
+
+  if (defaultImplPluginName != null) {
+    if (implementsPackage != null && implementsPackage.isNotEmpty) {
+      throwToolExit(
```

Comment:
For explaining that, we have to look into the [former implementation](https://github.com/flutter/flutter/blob/cc9ac7d13cccd24d72faab566f9fb59663c3c3d1/packages/flutter_tools/lib/src/flutter_plugins.dart#L1220). There, the case was not really considered. It was assumed, that if a package has an `implementsPackage`, it does not also have a `defaultImplPlugin`, so it was registered as implementation only, but did not reference a default plugin, which is wrong.

So we can allow this of course, but then I would also add / rewrite that test to support that behavior officially.

Reply:
I would propose opening an issue and link it in the error description, so one with a valid use case can further contribute to this. I think this was not explicitly part of the [design doc](https://docs.google.com/document/d/1LD7QjmzJZLCopUrFAAE98wOUQpjmguyGTN2wd_89Srs/edit#) and I would avoid enabling more features, which are beyond the scope of this PR.

Reply:
Ideally we would preserve the existing behavior rather than introduce a new hard error, but on the other hand since currently anything doing this is wrong, I guess I'm fine with a new error. If we get issues filed because some packages are (presumably accidentally) actually doing this, we can re-evaluate.

Reply:
I changed the error description to be more engaging, so the devs know, that this is not a hard limitation (685285f7a696261f4d670d96e79a2ed10ba27069)